### PR TITLE
MGMT-23733: add PublicIPPool CRD, controller, and feedback controller

### PR DESCRIPTION
**Pull request by akshaynadkarni**
## Summary

[MGMT-23733](https://redhat.atlassian.net/browse/MGMT-23733): Implement the
full PublicIPPool provisioning pipeline in osac-operator: CRD types, provisioning
controller, and feedback controller.

PublicIPPool represents a provider-managed pool of public IP addresses
(analogous to AWS BYOIP pools or Azure Public IP Prefixes). When a PublicIPPool
CR appears on the hub cluster, the provisioning controller triggers AAP to create
MetalLB IPAddressPool + L2Advertisement resources, and cleans them up on
deletion. The feedback controller watches the same CRs and reports phase and
capacity (total/allocated/available) back to the fulfillment service via gRPC.

**What's included:**

- **CRD types**: spec (CIDRs, IPFamily, ImplementationStrategy), status (Phase,
  Jobs, Conditions, capacity fields), CRD validation markers (MinItems=1 on
  CIDRs, Enum on IPFamily and ImplementationStrategy), int64 capacity fields
  for IPv6 support
- **Provisioning controller**: PublicIPPoolReconciler with handleUpdate/handleDelete,
  provisioning via RunProvisioningLifecycle, implementation-strategy annotation
  defaulting to metallb-l2, DeletionTimestamp guard on management-state check
- **Feedback controller**: PublicIPPoolFeedbackReconciler that syncs CR phase to
  fulfillment-service proto state (Progressing->PENDING, Ready->READY,
  Failed->FAILED, Deleting->DELETING) and capacity fields (total, allocated,
  available) via gRPC Update. Handles finalizer lifecycle, Signal RPC on
  last-finalizer removal, and graceful NotFound handling during deletion.
- **BSR tag bump**: v0.0.52 -> v0.0.53 to pull in PublicIPPool proto definitions
  from fulfillment-service PR osac-project/fulfillment-service#392
- **Registration**: both controllers registered in setupNetworkingControllers()
- **Unit tests**: 12 provisioning controller lifecycle tests, 11 CRD type tests,
  10 feedback controller tests (phase sync, capacity sync, finalizer
  add/remove, Signal, NotFound handling, label filtering)
- **Tenant test fix**: flaky DuplicateStorageClass event assertion fixed by
  switching from Expect(Receive()) to Eventually(Receive()) to handle stale
  events from reconcile retry loops

## Testing

```bash
cd osac-operator
go build ./...
go run github.com/onsi/ginkgo/v2/ginkgo run ./internal/controller/
# 273 of 273 Specs passed, 0 failed
```

### E2E: Provisioning Lifecycle (2026-04-14, tag mgmt-23733-260414-1415)

Deployed to edge22 with AAP playbooks from osac-project/osac-aap#238.

| Test | Result |
|------|--------|
| Empty CIDRs rejected at admission (MinItems=1) | PASS |
| Invalid IPFamily rejected at admission (Enum) | PASS |
| Invalid ImplementationStrategy rejected at admission (Enum) | PASS |
| Create PublicIPPool (full lifecycle: Progressing -> Ready) | PASS |
| Create with explicit metallb-l2 strategy | PASS |
| Create with omitted strategy (controller defaults) | PASS |
| Delete PublicIPPool (full lifecycle: Deleting -> removed) | PASS |

### E2E: Feedback Controller (2026-04-15, tag mgmt-23733-260415-1610)

Deployed to edge22. The fulfillment-service on edge22 does not yet have the
PublicIPPool gRPC server registered ([MGMT-23760](https://redhat.atlassian.net/browse/MGMT-23760) not implemented), so the
feedback controller gets `code = Unimplemented` when attempting gRPC calls.
This limits what can be verified end-to-end, but confirms the controller is
correctly wired and does not break the provisioning flow.

| Test | Result |
|------|--------|
| Operator startup with feedback controller (no errors) | PASS |
| Provisioning lifecycle without UUID label (feedback controller skips it) | PASS |
| Feedback controller reacts to UUID label (attempts gRPC, gets Unimplemented) | PASS |
| Deletion of UUID-labeled pool (no stuck finalizers) | PASS |

Full feedback flow (phase sync, capacity sync, Signal RPC) will be E2E-testable
once the fulfillment-service PublicIPPool gRPC server ([MGMT-23760](https://redhat.atlassian.net/browse/MGMT-23760)) is deployed.
These paths are covered by 10 unit tests with a mock gRPC server.

## Related PRs

- osac-project/osac-aap#238 (AAP playbooks and MetalLB L2 role)
- osac-project/fulfillment-service#392 (PublicIPPool proto definitions, merged)

## Ticket

[MGMT-23733](https://redhat.atlassian.net/browse/MGMT-23733)

<br>

<small>Assisted-by: Cursor/Claude</small>

### CODE REVIEW

**Comment by eranco74** [comment]
## Code Review: PublicIPPool CRD & Controller

### ✅ Overall Assessment
Solid controller implementation following established OSAC patterns. The reconciliation logic is clean, tests are comprehensive, and E2E verification on edge22 confirms the integration works end-to-end.

**Strengths:**
- Clean controller following existing patterns (SecurityGroup, Subnet)
- Proper finalizer handling with no stuck resource risk
- Config-version-based retry logic (avoids infinite loops)
- Comprehensive test coverage: 12 controller tests + 11 type tests (all passing)
- E2E verified on edge22 with AAP jobs 10186/10188
- Implementation-strategy annotation defaulting to `metallb-l2`

---

### 🔴 Critical Issue 1: Management-State Annotation Finalizer Leak

**File:** `internal/controller/publicippool_controller.go`  
**Lines:** 73-77

**Problem:** Early return when `management-state: unmanaged` skips deletion handling:
```go
val, exists := pool.Annotations[osacManagementStateAnnotation]
if exists && val == ManagementStateUnmanaged {
    log.Info("ignoring PublicIPPool due to management-state annotation", "management-state", val)
    return ctrl.Result{}, nil  // ❌ Skips handleDelete even during deletion!
}
```

**Impact:** Resources with `management-state: unmanaged` that get deleted will:
- Have stuck finalizers (CR never removed from etcd)
- Leak AAP resources (MetalLB IPAddressPool/L2Advertisement never cleaned up)

**Recommended Fix:**
```go
val, exists := pool.Annotations[osacManagementStateAnnotation]
if exists && val == ManagementStateUnmanaged && pool.ObjectMeta.DeletionTimestamp.IsZero() {
    log.Info("ignoring PublicIPPool due to management-state annotation", "management-state", val)
    return ctrl.Result{}, nil  // Only skip when NOT being deleted
}
```

**Verification:** Add a test case:
```go
It("should cleanup finalizer for unmanaged pool on delete", func() {
    pool.Annotations[osacManagementStateAnnotation] = ManagementStateUnmanaged
    pool.DeletionTimestamp = &metav1.Time{Time: time.Now()}
    // ... verify finalizer removed and handleDelete called
})
```

---

### 🔴 Critical Issue 2: Missing CRD Validation

**File:** `api/v1alpha1/publicippool_types.go`  
**Lines:** 84-90

**Problem:** No validation on critical immutable fields:
```go
CIDRs []string `json:"cidrs"`       // No MinItems, no format validation
IPFamily string `json:"ipFamily"`   // No enum restriction
```

**Impact:**
- Users can create pools with empty CIDR lists → AAP job fails with unclear error
- Users can set `IPFamily: "ipv5"` → validation happens at AAP runtime, not admission time

**Recommended Fix:**
```go
// CIDRs is the list of CIDR blocks for this pool. All CIDRs must match the declared IPFamily.
// +kubebuilder:validation:Required
// +kubebuilder:validation:MinItems=1
// +kubebuilder:validation:MaxItems=100
CIDRs []string `json:"cidrs"`

// IPFamily indicates the IP address family for this pool (IPv4 or IPv6)
// +kubebuilder:validation:Required
// +kubebuilder:validation:Enum=IPv4;IPv6
// +kubebuilder:validation:Type=string
IPFamily string `json:"ipFamily"`
```

After changes:
1. Run `make manifests` to regenerate CRD
2. Verify `config/crd/bases/osac.openshift.io_publicippools.yaml` has enum/minItems

**Note:** Kubernetes doesn't have built-in CIDR format validation in kubebuilder markers. Runtime CIDR validation would require a validating webhook (can be follow-up work).

---

### 🔴 Critical Issue 3: Capacity Field Type Mismatch

**File:** `api/v1alpha1/publicippool_types.go`  
**Lines:** 136-145

**Problem:** Status capacity fields use `int32` but should use `int64`:
```go
Total int32 `json:"total,omitempty"`      // Should be int64
Allocated int32 `json:"allocated,omitempty"`  // Should be int64
Available int32 `json:"available,omitempty"`  // Should be int64
```

**Impact:** Integer overflow for IPv6 pools (a /64 contains 2^64 addresses, exceeds int32 max of ~2.1 billion).

**Recommended Fix:**
```go
Total int64 `json:"total,omitempty"`
Allocated int64 `json:"allocated,omitempty"`
Available int64 `json:"available,omitempty"`
```

**Coordination:** This fix should be applied **after** fulfillment-service PR #392 merges with the same int64 change in the proto schema.

---

### 📋 Merge Coordination

1. **Wait for PR #392** to merge with int64 fix
2. **Update buf dependencies:** After #392 merges, CI will push a new BSR tag for `buf.build/osac-project/private-api`
3. **Apply all three fixes** above
4. Run `buf mod update && buf generate` to pull new proto schema
5. Update CRD types to int64 to match proto
6. Regenerate CRDs with `make manifests`

---

**Recommendation:** All three issues should be addressed before merge. They're straightforward fixes that prevent production bugs (finalizer leaks, validation gaps, overflow risks).

File: api/v1alpha1/publicippool_types.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	Region string `json:"region"`
+
+	// CIDRs is the list of CIDR blocks for this pool. All CIDRs must match the declared IPFamily.
+	// +kubebuilder:validation:Required
+	CIDRs []string `json:"cidrs"`
+
+	// IPFamily indicates the IP address family for this pool (IPv4 or IPv6)
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Type=string
+	IPFamily string `json:"ipFamily"`
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply api/v1alpha1/publicippool_types.go:34]: I have addressed this.

**Comment by akshaynadkarni** [comment]
## E2E Verification (2026-04-14)

Deployed to edge22 with tag `mgmt-23733-260414-1415`. All review feedback changes verified end-to-end.

### Tests Conducted

| # | Test | Result |
|---|------|--------|
| 1 | Empty CIDRs rejected at admission | PASS |
| 2 | Invalid IPFamily (`"ipv5"`) rejected at admission | PASS |
| 3 | Create PublicIPPool without Region field (full lifecycle: Progressing -> Ready, AAP job 10663) | PASS |
| 4 | Delete PublicIPPool (full lifecycle: Ready -> Deleting -> CR removed) | PASS |

### CRD Validation Output

```
# Empty CIDRs
The PublicIPPool "test-bad-empty-cidrs" is invalid:
  spec.cidrs: Invalid value: 0: spec.cidrs in body should have at least 1 items

# Invalid IPFamily
The PublicIPPool "test-bad-ipfamily" is invalid:
  spec.ipFamily: Unsupported value: "ipv5": supported values: "IPv4", "IPv6"
```

### Changes Verified

- **CRD validation**: `MinItems=1` on CIDRs, `Enum=IPv4;IPv6` on IPFamily
- **Region removal**: CR accepted without region, printer columns show IPFamily (no Region)
- **int32 -> int64**: CRD applied cleanly (behavioral verification deferred to feedback controller)
- **DeletionTimestamp guard**: verified via unit test (edge case requires manual annotation mid-lifecycle)
- **AAP changes** (osac-aap PR #238): project synced, create/delete flows succeeded with updated playbooks

<small>Assisted-by: Cursor/Claude</small>

File: internal/controller/publicippool_controller.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+				Timestamp:              metav1.NewTime(time.Now().UTC()),
+				State:                  v1alpha1.JobStatePending,
+				Message:                "Deprovisioning job triggered",
+				BlockDeletionOnFailure: result.BlockDeletionOnFailure,
+			}
+			pool.Status.Jobs = provisioning.AppendJob(pool.Status.Jobs, newJob, r.getMaxJobHistory())
+			log.Info("deprovisioning job triggered", "jobID", result.JobID)
+			return ctrl.Result{RequeueAfter: r.getStatusPollInterval()}, nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/controller/publicippool_controller.go:268]: Keeping as-is for consistency with existing controllers (VirtualNetwork, Subnet, SecurityGroup all follow the same pattern).

**Review by eranco74 — APPROVED** [review]
## ✅ LGTM

Comprehensive PublicIPPool implementation with solid test coverage and E2E verification on edge22.

### CodeRabbit Issues Resolved
- ✅ CRD validation markers added (Enum on ImplementationStrategy)
- ✅ Management-state check moved above deletion logic
- ✅ All nitpicks acknowledged (redundant GetName, dead code, etc.)

### Highlights
- CRD schema matches proto (int32 capacity fields, phase enum, Jobs tracking)
- Config-version-based retry logic with exponential backoff
- Implementation-strategy annotation auto-populated (defaults to metallb-l2)
- Finalizer lifecycle correct
- 33 total tests across provisioning controller, CRD types, feedback controller
- E2E tested: CRD admission, full lifecycle (Progressing → Ready → Deleting)

### One Minor Recommendation
The management-state check at line 813 should include a DeletionTimestamp guard to prevent finalizer leaks if a pool is marked Unmanaged after provisioning:

```go
if exists && val == ManagementStateUnmanaged && pool.ObjectMeta.DeletionTimestamp.IsZero() {
    return ctrl.Result{}, nil
}
```

This prevents orphaned resources if a user marks a provisioned pool as Unmanaged then deletes it. Low risk for v7.0 (no AAP Direct yet), but worth fixing before AAP Direct support.

/lgtm
